Fix rendering of allfaces nodes with transparent textures

x2048 opened the pull request:
This PR fixes #11119.

I contains the following changes:

* Mapblock meshes are generated 'towards the player' with dynamic precedence of axes.
* When the player moved, mapblocks are scheduled for mesh regeneration as required.
* Mapblock draw list is sorted by distance to the player.

## To do

This PR is Ready for Review.

What could be done:

- [x] Fix rendering of overlapping different transparent materials
- [x] Improve markBlocksDirty to avoid remeshing solid blocks changed by shaders
- [x] Transparent nodes placed in water are not rendered correctly
- [x] Side-by-side different transparent materials still produce artifacts (see comments)
- [x] Add more test nodes to devtest (allfaces and check if multi-faced liquid is there)

## How to test

1. Register a node with this definition:
```
{
    description = "Test",
    drawtype = "allfaces",
    tiles = {"white.png^[opacity:10"},
    inventory_image = "white.png^[opacity:129",
    use_texture_alpha = "blend",
    sunlight_propagates = true,
    walkable = false,
    pointable = false,
    diggable = false,
    floodable = true,
    paramtype = "light",
    air_equivalent = true
}
```
Where `white.png` is a single white pixel.

2. Create a cloud of these nodes.
3. Walk around and inside the cloud

*Expected behavior:* Cloud looks uniform from all directions, there are no significant visual artifacts when moving inside the cloud.

appgurueu commented:
Could this mostly fix our transparency sorting issues? What is the impact on performance?

Wuzzy2 commented:
Doesn't work:

https://satoshiupload.com/images/MGYwFGjYAe.png

I used your test node, but increased the opacity to make it more visible.
I have this missing face sometimes when I look at a certain angle, but the missing face might re-appear seemingly randomly when I move further. It happens very quickly.

Also, the inventory image is invisible, it's also invisible when I drop the item on the floor.

I also changed the alpha test nodes from DevTest from `"glasslike"` type to `"allfaces"`. At first it looks fine:
https://satoshiupload.com/images/pBCZiCERnN.png

But if you move around a bit, the nodes in the back disappear:
https://satoshiupload.com/i/xwg1golld0

x2048 commented:
@Wuzzy2, thanks for the thorough tests.

The reason you see disappearing and reappearing faces is that I implemented a variant of painter's algorithm in MapblockMeshGenerator::generate, so correct rendering depends on generating correct mesh. When you cross node boundary, the order must change, so I trigger remesh. This in turn happens asynchronously, and I need to touch many mapblocks, so there's some flicker, which I mask as much as possible by e.g. choosing priority of X and Z axes for the loops. The picture should stay consistent when you are not moving, let me know if you experience flicker while standing still and rotating camera.

This might be a completely naive implementation using the components I had at hand, so ideas to improve are welcome. So far I have two ideas:
* remesh player's current mapblock and the next in the look direction in the main thread. This will cause serious impact on overall game performance.
* look into irrlicht's MeshBuffers and reorder faces there by their proximity to the camera, again, in the main thread, and gain for the two closest blocks only.

As for the second test, I'll take a look at the alpha test nodes from devtest to see if I can reproduce what you see.

x2048 commented:
@appgurueu, the primary performance impact is on the mesh generation thread due to the continuous remeshing currently necessary to get the correct order of faces to be rendered. This is triggered on crossing node boundary by [ClientMap::markBlocksDirty](https://github.com/minetest/minetest/pull/11130/files#diff-338620f4b79e1a00e394bb90133eb962a50bfa461512d619452e879a38ae8bacR646).

Without shaders:
* When no transparent nodes in sight, it triggers 0-1 mapblocks per call at up to 20 calls per second (when moving Fast+Fly).
* When moving across a cloud of 5x2x4 mapblocks of transparent nodes along the Z axis, it triggers about 30 mapblocks per call

With tonemapping shader:

* When no transparent nodes in sight, it triggers about 40 mapblocks per call at 20 calls / second.
* When moving across a cloud of 5x2x4 mapblocks of transparent nodes, it triggers about 60 mapblocks per call.

The difference is due to the shader modifying the material, and I don't have a good check for transparent materials in [hasTransparency](https://github.com/minetest/minetest/pull/11130/files#diff-338620f4b79e1a00e394bb90133eb962a50bfa461512d619452e879a38ae8bacR638-R639). I'd appreciate advice on what to look at.

x2048 commented:
@Wuzzy2 Fixing inventory image was not my goal with this pull request. Would you expect it to be fixed here?

x2048 commented:
@Wuzzy2, I can confirm the problem you showed with devtest alpha nodes, looking into it.

Wuzzy2 commented:
> @Wuzzy2 Fixing inventory image was not my goal with this pull request. Would you expect it to be fixed here?

No.

appgurueu commented:
My thoughts:

1. Nodes have fields telling you whether they can have semitransparency (see https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L7362-L7373). You can use this to determine whether a mapblock has semitransparency by simply looping over all nodes. That value can be cached and updated as the mapblock is updated/replaced.

2. Vertices, texture coordinates, and normals are all the same, no matter how you view a mapblock. These can be generated once and cached. What does change however is the draw order of triangles as specified by the indices. Suggestion: Recalculate just this order, or, consider generating "all" (from all look directions) possible orders and caching them (might be too memory-intensive).

3. Do the same for entities: Sort them like the mapblocks before drawing them. Use https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L4376 to determine whether an entity supports semitransparency.

4. Nontrivial: Do the same for entity & node meshes: Provide different indices for different viewing angles. Again, consider recalculating vs. caching.

5. Nontrivial: Entity-mapblock interaction: Sort all scene elements (entities & mapblocks) properly, draw them in order. If necessary, render part of a mapblock behind, part before the entity.

Alternatively:

1. Implement order-independent transparency, such as an A-buffer (extended depth buffer storing multiple fragments). Most likely won't work well with OpenGL or be quite limited and negatively impact performance.

If either of this methods is done completely, #95 and related semitransparency issues could be finally closed.

x2048 commented:
@Wuzzy2, I've looked at the disappearing and reappearing nodes. What I see is that each TileLayer becomes a separate mesh created on demand via MeshCollector, and depending on which mesh is created first, the nodes in the front will either overwrite or blend with the nodes in the back. Without the fix the behavior was that certain look angles (e.g. -X) produced correct blending while other angles (e.g. +X) produced 'missing' nodes just because the loops were always going in +X direction.

I think, simply grouping PreMeshBuffers by TileLayer is not enough, and at least distance to the player must be taken into account. I did not think to much about how to solve this yet, but I have MapblockMeshGenerator::generator now draws nodes by descending Manhattan distance to the player, which may be handy.

x2048 commented:
Thanks @appgurueu 

1. Indeed, that's the API we are using for testing :) I'll see if I can use it.
2. Sorting could be an option. Is it possible to manipulate SMesh/SMeshBuffer after it has been created?
3. 4. 5. May I wonder which problems with entities we are trying to solve? I quickly tried placing a bunch of zombies in a fog, and it looked fine.

A-buffer looks quite complicated, and it does not seem like it would be an easy task with Irrlicht, right?

x2048 commented:
Hey @Wuzzy, I've hacked MeshCollector and ClientMap to avoid blind sorting by materials when rendering, so various transparent blocks are now rendered correctly from all angles ([image](https://satoshiupload.com/images/587mcReRCI.png)).

@appgurueu I've made another performance test on an MCL2 scene, where I added a 3x1x6 mapblocks cloud of fog on a field of grass. Duration of transparent pass increased from 0.4ms to 4ms on Ryzen 7 4800H compared to 5.4.0, and number of transparent drawcalls grew from 100 to 1000.

Wuzzy2 commented:
This is great! I very much appreciate your work on this. I have tested it, and the depth sorting of all allfaces and glasslike nodes seems to work now as expected. Great! I always see the correct color now, now block disappears. I have tested it with the colored glass from MineClone 2 and it works! No matter how hard I try, no more breakage.

Also, the blinkenlights bug with translucent ice from MCL2 seems to be fixed as well (see #95). (translucent ice must be manually enabled in MCL2 with a setting first). Translucent ice no longer flashes around.

However, depth sorting still is bugged with liquids: https://satoshiupload.com/images/Cm8gm7e49J.png

In DevTest, you have various liquid drawtype test nodes to play around with. Use those for testing, if you plan to fix liquids as well.

I'm just telling this for your information. I think this PR is already perfectly mergable. This is BIG news for Minetest!

If you can't fix liquids right now, that's fine. The bugfix is already a huge improvement. Liquids can always be fixed in a later PR.

x2048 commented:
Thank you @Wuzzy2, I really appreciate your feedback, and I'm glad to hear that some old bugs go away with the patch.

I've looked at the liquid sources, and it looks like I can fix them by moving from updateAllFastFaceRows to MapblockMeshGenerator, which has much of the code already implemented for the flowing liquids. I'll take a look within a coupld of days.

I also spotted a bug with ordering of the nodes in a mapblock when different transparent materials are placed side-by-side. I reproduce it with a 3x3 grid of alpha_128 with a texture_alpha_128 in the center:
```
  A A A
  A B A
  A A A
```
When moving round this square, the sides of B are sometimes masked by the closest A, I'll see if I can fix is quickly enough. I am tracking progress in the issue description.

x2048 commented:
@Wuzzy2, I've unified rendering of still and flowing water, which fixes rendering of ice, underwater glass etc. The performance penalty is that a small sea in MCL2 results in 3K+ drawcalls in transparent pass when looked at from above.

appgurueu commented:
>     2. Sorting could be an option. Is it possible to manipulate SMesh/SMeshBuffer after it has been created?

I don't know. [Perhaps](http://irrlicht.sourceforge.net/docu/classirr_1_1scene_1_1_i_mesh_buffer.html#a76c0013378012af7aeb6cb8f4ea8f9a1). If not, we could now change Irrlicht, as it has been forked.

>     3. , 4.  and 5. May I wonder which problems with entities we are trying to solve? I quickly tried placing a bunch of zombies in a fog, and it looked fine.

Entities appear to already be sorted that way, regardless of transparency, so 3 might be irrelevant.

Try using semitransparent mesh nodes, or semitransparent entities, and observe that overlapping transparent faces don't combine well (4).

5, if semitransparent entity faces and node faces combine colors, it will look wrong in some cases.

Your should probably ignore 4. and 5. for this PR, it is out of scope and not that much of an issue. It's more of a rough overview of what would be needed in order to completely fix all issues with semitransparency.

> A-buffer looks quite complicated, and it does not seem like it would be an easy task with Irrlicht, right?

Indeed. There also is no proper hardware support yet.

Wuzzy2 commented:
Nope, the latest changes broke the rendering of liquid source nodes somehow. In MineClone 2, the side faces of water sources no longer renders. This is a regression.

x2048 commented:
What is your scenario? I disabled rendering of side and top faces when the neighbor has visual solidness 1 or is an ignore block, but they still render next to e.g. air.

Wuzzy2 commented:
In MineClone 2, place a water source on the floor, and then put 4 signs around it, so the water doesn't flow away. Now the side faces are invisible, only the top/bottom faces are.

This worked before, even within your PR, so I suspect this must have become broken recently.

x2048 commented:
I have fixed rendering of liquid sources by reusing the first normal tile for both top/bottom and sides. I've also spent some time improving performance, and now 'layered' rendering applies to transparent nodes only. Current performance is around 80% of 5.4.0.

Wuzzy2 commented:
> I have fixed rendering of liquid sources by reusing the first normal tile for both top/bottom and sides.

I believe this is still wrong. According to my tests and looking at the code, I have concluded that nodes of drawtype `liquid` just use `tiles` like nodes of drawtype `normal`. You could theoretically define all 6 tiles differently, but this is rarely done in practice. But also the “fill-in missing textures” rules of `tiles` applies here. E.g. if you only define tiles 1 and 2, all missing tiles will be copies of the last tile, in this case tile 2. Just like for `normal`-drawtyped nodes.

That having said, here's what I've found:

https://satoshiupload.com/images/yIXIAJIPaT.png

Weird corner rendering of 2×2 water in MCL2. You can also reproduce with `testnodes:alpha_191` and `basenodes:water_source` in DevTest. If you move around, the corner might also flash.

https://satoshiupload.com/images/8qTPwbUQXy.png

If you build a tower of transparency test nodes and just put a water source on top, this happens. This is about water flowing downwards. These horizontal planes of water are only visible from the outside, not if you are inside the water.


https://satoshiupload.com/images/DCGmiCKJ03.png

Not sure if intentional, but if you look at the sides of translucent ice in the water, you see a water face, not an ice face. Note: If water is bordering sand or any other non-translucent node, the face of that node (e.g. sand) is rendered, not the water face.
Same in DevTest. Maybe this is because of `allfaces` which is usually meant for leaves? Might be OK to not change this, I do not think this is REALLY related to this PR (since it technically is not transparency sorting anymore), nor do I think this is REALLY a bug, even. This needs to be thought through first IMO.


So, liquids still make some problems, but it seems you're getting closer! Keep it up! :D

I have a suggestion: Consider adding more transparency test nodes into DevTest (in `testnodes` mod) to test out more esoteric cases. I think we are lacking one major drawtype for translucency tests (either `allfaces` or `glasslike`).

x2048 commented:
@Wuzzy2 I took another approach to rendering liquids:
* Threw away all the previous code
* Anything solid is rendered by fast faces code (including liquids). This brings back performance on slower machines.
* Transparent liquid is now rendered by a new method drawLiquidSourceNode() that reproduces all the rules from fast faces.
* Flowing liquid avoids creating side faces when adjacent to glass/allfaces or similar.
* Two flowing liquids will still create faces towards each other, with a small offset to avoid z-fighting

This should fix the problem with e.g. ice and use of different textures in water sources.

The problem with the tower of flowing water seems to be caused by lack of depth sorting in liquid_base_faces. What appears as possible horizontal faces is caused by incorrectly overlapped side faces. I don't think I can fix it now.

The 2x2 water source rendering in mineclone seems to be caused by some ordering issue when generating mapblock mesh. I am still looking into this.

x2048 commented:
Hi @Wuzzy2, @appgurueu, @sfan5 , I've taken another look at the things and added depth sorting to liquid sources. I also fixed a couple of glitches with the order of rendering and added a few optimizations.

I did not touch rendering of flowing water because it's quite complicated, and would require significant refactoring on top of what's already here.

@appgurueu I did not look into SMesh here, because blind vertex sorting will work correctly in all cases due to how I generate transparent meshes as 'layers' of nodes based on viewpoint.

@sfan5 May I ask you to run this through OpenGL profiler you were using to see if I make good use of Z-buffer now?

I think all major work is done for this PR and would like to hear if there are any blockers to merging it in.

Thank you for your help.

appgurueu commented:
> @appgurueu I did not look into SMesh here, because blind vertex sorting will work correctly in all cases due to how I generate transparent meshes as 'layers' of nodes based on viewpoint.

That is completely fine. This should basically fix/introduce Z-sorting at a mapblock mesh generation level, and by rendering the mapblocks in the proper order?

> I did not touch rendering of flowing water because it's quite complicated, and would require significant refactoring on top of what's already here.

I don't think it's that complicated. You still have side & top faces; you "only" have to do is properly reorder them depending on look direction.

Anyways, this PR already is awesome work!

x2048 commented:
Thanks, @appgurueu. 

> This should basically fix/introduce Z-sorting at a mapblock mesh generation level, and by rendering the mapblocks in the proper order?

Yes, MapblockMeshGenerator::generate() now builds Z-sorted layers of meshes based on manhattan distance (voxel worlds are great!) from the camera.

As for flowing water rendering, I get the principle, but it's spread across 5 methods, 3 of which render various faces. Needs some refactoring in a new PR.

appgurueu reviewed (approved):
I can hardly judge the code style, but conceptually this should be fine. Flowingliquids might be left to a follow-up PR.

Wuzzy2 commented:
I can accept if you don't fix all the transparency bugs at once. At some point, it's better to just accept the PR as-is, leaving the remaining bugfixes for later.
However, this PR should still be 100% regression-free.
And I have found several regressions:

1. Liquids are now invisible behind a "classic" allfaces node like leaves. Surround water by leaves and look at the leaves from outside. The water is invisible now. Here's what I suggest: When a liquid source borders an allfaces node, render the face when viewed from the side of the allfaces node, but do not render the face from the liquid's point of view. That way, liquids don't become invisible behind leaves but it would still fix the MCL2 ice bug
2. Liquid sources now always render their backfaces even if `backface_culling` is not set. But as far I know, backfaces must never be rendered for any drawtype unless `backface_culling=false` was specified explicitly (I'm not sure, ask a coredev).
3. I have not found more regressions, but please make sure that previous reasonable behavior is not acccidentally destroyed.

> Two flowing liquids will still create faces towards each other, with a small offset to avoid z-fighting

As far I know, offsets are almost *never* a good idea to fix z-fighting, unless the offset is actually intended as a feature and large enough. IIRC Irrlicht/IrrlichtMt provides a function to deal with z-fighting (ask a coredev).
But I think in the long run, all bordering [flowing]liquids of the same "family" (source+flowing, as specified in `liquid_alternative_*`) should never have faces between each other, regardless of `backface_culling`. That's so you can see through inside the node, just like in a real liquid. But this is kinda off-topic here, you can safely ignore this for your PR.


> The problem with the tower of flowing water seems to be caused by lack of depth sorting in liquid_base_faces. What appears as possible horizontal faces is caused by incorrectly overlapped side faces. I don't think I can fix it now.

I think it's fine as long it is not a regression.

Finally, I request some DevTest cleanup:

- Remove `inventory_image`/`wield_image` from the new test nodes. The drawtype test nodes intentionally do NOT have these things set so the default node inventory rendering is used. Please do this even if the items would look wrong in the inventory now, that's kind of the point (to expose bugs with the inventory rendering). Don't bother about inventory render fails, this is off-topic.
- Make the 6-face liquid non-walkable

appgurueu commented:
> As far I know, offsets are almost _never_ a good idea to fix z-fighting, unless the offset is actually intended as a feature and large enough. IIRC Irrlicht/IrrlichtMt provides a function to deal with z-fighting (ask a coredev).

No, offsets are the only means known to me to fix Z-fighting. I doubt Irrlicht can provide anything better than that.

x2048 commented:
@Wuzzy, thanks for the feedback. 

I've now tested many combinations of leaves, glass and various liquids in devtest and made transparent rendering behave the same way as opaque rendering when if comes to generating faces and forcing backface culling. I've also cleaned up the nodes I added to devtest and aligned the textures on 6face liquid and 6face test node.

Adding support for 6face liquids introduced a performance problem when rendering seas (200K+ transparent drawcalls in MCL2 when approaching a small sea) since every rendered face of every visible water node produced a unique mesh. I've optimized that out now for almost all of cases (checked with MCL2, NodeCore, Whynot), but rendering vast sheets of transparent ice on top of transparent water is still slow (I saw 20K+ drawcalls in the reference scene, see below). I don't have a good solution yet, and since transparent ice is currently experimental, I don't think I'll fix it in this PR.

My reference scene is: MCL2, mapgen: flat, seed: 1024, no caverns, pos: (1095, 2, -655), view distance: 540, look +Z.

Wuzzy2 commented:
So I noticed the following bugs:

1) In MCL2, the side of flowing water has turned green, lol
2) If you put a flowing water node next to a water source node of a different type (e.g. flowing water vs river water source), there's z-fighting

x2048 commented:
Thank you, @Wuzzy2.
I cannot confirm 1. with the latest MCL2. Are you using a specific texturepack?
Regarding 2., Minetest 5.4.0 behaves exactly the same way in case two liquids collide, or when liquid flows next to leaves or glass, so it does not look like a regression. I'll be happy to fix this in the next PR (dedicated to liquids).

Wuzzy2 commented:
Both my previous complaints are actually invalid. The first one was because I had used old irrelevant testing code (whoops!). And yeah, ignore the 2nd one, it's not a regression so it's OK to leave it untouched for now.

I did new testing now and it seems to mostly work now, with two exceptions:

1) In DevTest, if you place the red alpha test node `testnodes:allfaces_alpha_191` and build a floor with it and then look down and then walk around to look at the textures, you will sometimes see the side faces be invisible for a split seconds as you move.

2) In MCL2 If you put water on top of the translucent ice and slowly walk on it and, you will sometimes also see some water faces flash in and out of existence very quickly

Those two bugs are a bit subtle but still irritating.

Areas in which I FAILED to find errors:
- Translucent allfaces nodes vs translucent allfaces nodes: No node "disappears" (but see bug 1 above)
- Translucent allfaces nodes vs translucent glasslike nodes: No node "disappears"
- Translucent glasslike nodes vs translucent glasslike nodes: No node "disappears"
- Colored glass vs colored glass in MCL2: No node "disappears"
- Edges of translucent ice in MCL2: looks correct now
- Water vs leaves: Proper rendering from both sides
- Regular allfaces nodes
- Water and flowing water: No surprising faces when inside the liquid
- Liquid vs liquid: OK (except the z-fight mentioned above which doesn't count)
- Bug #9324 no longer exists

I think I tested pretty much everything that could be tested. The most important news of this PR is that colored translucent glass will finally work properly.

https://satoshiupload.com/images/PRbaRhSNaO.png

This looks very good now. I think if you manage to fix the 2 remaining bugs, this PR is probably ready for merge, but you still need core dev approval (I am NOT a core dev). Impressive work so far! :D

----

By the way, since you plan to work on liquid rendering in the future, here are some issues you want to pay attention to (in future PRs, of course):
#9276, #9323, #5780, #3075

Wuzzy2 commented:
Minor DevTest change request: `^[colorize` the sides of the 6faces liquid node in blue to make it more visible in the inventory. I literally could not see this node in the inventory for over an hour.

x2048 commented:
Thank you for thorough testing and detailed feedback.

Flickering in both cases happens because correct rendering depends on player's position, and mesh is regenerated in a separate thread. Making it synchronous (even on demand) will severely impact FPS and overall gameplay experience, so I cannot do this without drastic changes to the rendering pipeline. The best I can do so far is to try to play with the order of nodes so that the flickering is reduced in major cases. Do you think this is a blocker for inclusion into the mainstream?

> Minor DevTest change request: ^[colorize the sides of the 6faces liquid node

Done.

sfan5 commented:
If this pull request is ready for review please mark it that way using Github's interface.

x2048 commented:
@sfan5 Done

Desour commented:
In `markBlocksDirty`, I don't think you always must remesh all mapblocks that have the same blockpos as the current or previous block on _any_ axis. If you didn't change position in one axis (eg. you moved on flat ground) then you can ignore blocks where the pos is only the same on this axis (one of the 3 axis-aligned 2D mapblock planes that go through you, without the 1D lines). And if you move along one axis (eg. you fall down a fall shaft) you can ignore everything but the blocks that have the same pos component of that axis (you only have to remesh the one 2D plane, instead of all 6).
I hope you can understand what I mean here. If not I'll provide visual material.

Also in `markBlocksDirty` you could make use of sectors being Y-wise stacks of mapblocks and move some checks into outer loops.

(Btw., just in case you didn't find this already, the code style is documented: https://dev.minetest.net/Code_style_guidelines (much old code doesn't use it though).)

(I didn't look through the whole code yet.)

Wuzzy2 commented:
> Flickering in both cases happens because correct rendering depends on player's position, and mesh is regenerated in a separate thread. Making it synchronous (even on demand) will severely impact FPS and overall gameplay experience, so I cannot do this without drastic changes to the rendering pipeline. The best I can do so far is to try to play with the order of nodes so that the flickering is reduced in major cases. Do you think this is a blocker for inclusion into the mainstream?

I do not. The issue at hand doesn't happen often. Your PR overall is still a massive improvement.

x2048 commented:
Thank you, @Desour. I've now optimized markBlockDirty based on relative movement of the player across the nodes. I still need to remesh blocks in all three planes (XY, XZ, YZ) if player's position changes, because rendering of transparent nodes depends on the player's node position, not only the mapblock the player is in.

Thanks for sharing the coding style, I've now made corrections where I stepped away from it.

@Wuzzy2, I reverted changes related to #9324 because they broke rendering of flowing liquids in other cases, for example water flowing in tree leaves would be invisible from below and the sides.

Andrey2470T commented:
Nice to see the oldest visual bug in MT is going to be fixed! Could this PR potentially also fix #10218 ?

Wuzzy2 commented:
This PR still is looking good, after some tests.

x2048 commented:
@Andrey2470T The patch fixes #10218, but introduces flickering when player moves around "simple window" tiles: tile edges are rendered thick or thin from move to move. I'll need to take a look at this.

sfan5 commented on src/client/content_mapblock.cpp:
```diff
@@ -1440,19 +1548,112 @@ void MapblockMeshGenerator::drawNode()
+		collector->startNewMeshLayer();
+		for (s16 i = 7; i >= 0; --i) {                        // loop over octants
+
+			struct octant *octant = octants + i;
+			struct boundary *boundary = bounds + i;
```
```suggestion
			const octant &octant = octants[i];
			const boundary &boundary = bounds[i];
```
would be neater IMO

x2048 replied:
Done

sfan5 commented on src/client/mapblock_mesh.cpp:
```diff
@@ -664,19 +664,22 @@ static u8 face_contents(content_t m1, content_t m2, bool *equivalent,
 
 	if (c1 > c2)
-		return 1;
+		return f1.drawtype == NDT_NORMAL || f1.drawtype == NDT_PLANTLIKE_ROOTED || f1.alpha == ALPHAMODE_OPAQUE ? 1 : 0;
```
This looks like the wrong approach to me, `nodedef.cpp` is supposed to set solidness so that MapBlockMesh _doesn't_ have to do these comparisons.

x2048 replied:
We need a new system for solidness weights then, what is the best place to discuss it?

x2048 replied:
I've implemented new levels of solidness and removed visual_solidness entirely.

sfan5 commented on src/client/meshgen/collector.cpp:
```diff
@@ -95,10 +96,47 @@ PreMeshBuffer &MeshCollector::findBuffer(
+
+	// avoid scanning the entire list of buffers when filling with the same material
+	if (latest_buffers[layernum].size() > 0 && 
+			buffers[latest_buffers[layernum].back()].layer == layer && 
```
Am I missing something or can you just replace that by `buffers.back()`?

x2048 replied:
I've rewritten the code without using vectors of integers, much simpler now.

sfan5 commented on src/client/meshgen/collector.cpp:
```diff
@@ -95,10 +96,47 @@ PreMeshBuffer &MeshCollector::findBuffer(
+		if (allow_reuse && latest_buffers[tile_layer].size() > 0)
+			latest_buffer = latest_buffers[tile_layer].back();
+
+		for (s16 index: latest_buffers[tile_layer])
```
This code should have comments to explain its' purpose and use.

x2048 replied:
I've now simplified the method and added comments.

sfan5 reviewed (commented):
Haven't tested yet but here's what caught my eye reading the code.

x2048 commented on src/client/clientmap.h:
```diff
@@ -127,12 +141,34 @@ class ClientMap : public Map, public scene::ISceneNode
+	public:
+		MapBlockComparer(const v3s16 &camera_block) : m_camera_block(camera_block) {}
+
+		MapBlockComparer(const MapBlockComparer& origin) : m_camera_block(origin.m_camera_block) {}
```
Copy constructor is used internally by std::map<>. I can add annotate with a comment for clarity.

x2048 replied:
Indeed. Removed the constructor.

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -608,3 +634,88 @@ void ClientMap::PrintInfo(std::ostream &out)
+					++blocks_on_axis;
+
+					// if there is known transparency
+					auto block_distance = abs(block_pos.X - current_block.X) +
```
Correct.

x2048 replied:
I've added an inline function to irr_v3d.h and used place throughout the code.

x2048 commented on src/client/clientmap.h:
```diff
@@ -86,10 +86,24 @@ class ClientMap : public Map, public scene::ISceneNode
+
+		// reorder the blocks when camera crosses block boundary
+		if (previous_block != current_block)
+			updateDrawList();
```
updateDrawList is called roughly 5 times a second, see [game.cpp#3833](
https://github.com/minetest/minetest/blob/a24899bf2dcd58916922d671ee8761448b6876e5/src/client/game.cpp#L3833).

I need to force-update draw list when a block boundary is crossed to reorder the blocks by new Manhattan distances and avoid flicker. It may be possible to change the condition in game.cpp, but I found this call easier.

x2048 replied:
I've now replaced this call with a flag for Game to call updateDrawList from the game loop.

x2048 commented:
@Andrey2470T, I've pushed a fix for the flicker mentioned earlier, so this PR now fully fixes #10218.

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -608,3 +634,88 @@ void ClientMap::PrintInfo(std::ostream &out)
+
+			// Skip sectors which are not on axis or where there was no movement between nodes
+			bool same_x = sp.X == pointer->X && current_node.X != previous_node.X;
+			bool same_z = sp.Y == pointer->Z && current_node.Z != previous_node.Z;
```
No it can't for this loop, because it depends on the loop variable (pointer).

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -408,13 +417,30 @@ void ClientMap::renderMap(video::IVideoDriver* driver, s32 pass)
 		}
 	}
+
+	drawcall_count += transparent_buffers.size();
```
Done

x2048 commented on src/client/clientmap.cpp:
```diff
@@ -608,3 +634,88 @@ void ClientMap::PrintInfo(std::ostream &out)
+							abs(block_pos.Y - current_block.Y) +
+							abs(block_pos.Z - current_block.Z);
+
+					if (block_distance < 10) {
```
Done

x2048 commented on src/client/clientmap.h:
```diff
@@ -86,10 +86,24 @@ class ClientMap : public Map, public scene::ISceneNode
+		// reorder the blocks when camera crosses block boundary
+		if (previous_block != current_block)
+			updateDrawList();
 	}
```
Done

x2048 commented on src/client/content_mapblock.cpp:
```diff
@@ -1440,19 +1548,112 @@ void MapblockMeshGenerator::drawNode()
+	player_pos = floatToInt(data->m_client->getCamera()->getPosition(), BS);
+
+	v3s16 player = v3s16(
+		std::max<s16>(-1, std::min<s16>(MAP_BLOCKSIZE, player_pos.X - blockpos_nodes.X)),
```
Done

Wuzzy2 commented:
Typo:

>  	SOLIDNESS_AIR,		// air and irlike nodes

x2048 commented on src/client/content_mapblock.cpp:
```diff
@@ -1440,19 +1548,112 @@ void MapblockMeshGenerator::drawNode()
-		n = data->m_vmanip.getNodeNoEx(blockpos_nodes + p);
-		f = &nodedef->get(n);
-		drawNode();
+	player_pos = floatToInt(data->m_client->getCamera()->getPosition(), BS);
```
Changed now to use QueuedMeshUpdate and MeshMakeData to convey camera position.

x2048 commented:
@Wuzzy2  Thanks for caching. Fixed now.

@sfan5 Thank you for the review, I've now resolved all the comments. Looking forward to hearing your feedback in the second round.

sfan5 reviewed (commented):
Just a few minor comments. I'll actually test the PR next time.

x2048 commented on src/client/content_mapblock.h:
```diff
@@ -81,6 +81,7 @@ class MapblockMeshGenerator
 	video::SColor color;
 	TileSpec tile;
 	float scale;
+	v3s16 player_pos;           // player's position at the time of mesh generation
```
Done

x2048 commented on src/client/content_mapblock.cpp:
```diff
@@ -1440,19 +1551,110 @@ void MapblockMeshGenerator::drawNode()
+		for (s16 i = 7; i >= 0; --i) {                        // loop over octants
+
+			struct octant &octant = octants[i];
+			struct boundary &boundary = bounds[i];
```
Done

x2048 commented on src/client/mapblock_mesh.h:
```diff
@@ -125,6 +132,11 @@ class MapBlockMesh
 			m_animation_force_timer--;
 	}
 
+	inline bool hasTransparency()
```
Done

x2048 commented:
Hey @sfan5,

I've now implemented the suggestions, see 9dd44f9.

x2048 commented:
Hey @kilbith @nerzhul @lhofhansl @rubenwardy,

Can some of you take a look at this PR to see what's needed to get it merged?

Thank you.

rubenwardy reviewed (commented):
I don't feel comfortable with approving this PR, as it's way over my head. So I'll just do general code review and testing.

I needed to add `#include <memory>` to clientiface.h to fix a compile error finding `std::shared_ptr`

I've noticed that it fixes the issue with allfaces nodes, but there's a slight lag when walking:

https://user-images.githubusercontent.com/2122943/122651348-1ec82280-d130-11eb-96b5-5d4de2260f40.mp4

I'm on Manjaro linux (Dell XPS 13 9370, i7-8550U, Intel UHD Graphics 620)

x2048 commented:
@numberZero. can you provide feedback for this PR? @rubenwardy  recommended you as a reviewer.

Thank you.

x2048 commented:
@rubenwardy 
>     * I've noticed that it fixes the issue with allfaces nodes, but there's a slight lag when walking:       
>       
>       I'm on Manjaro linux (Dell XPS 13 9370, i7-8550U, Intel UHD Graphics 620)

The lag is there because transparency sorting requires remesh, which is currently done in a different thread. I can work on remesh if that's a blocker, but I'd prefer to do it in a separate PR.

hecktest commented:
> * Mapblock meshes are generated 'towards the player' with dynamic precedence of axes.
> 
> * When the player moved, mapblocks are scheduled for mesh regeneration as required.

This sounds expensive, wouldn't it be better to just reorder the indices of an existing mesh rather than remesh a whole block?
Also, is this only done for blockmeshes with an alpha blended material? I'll go through the code in a moment, but feel free to clarify in the meantime.

hecktest commented:
The PR breaks world-aligned water:
![bug](https://user-images.githubusercontent.com/42101236/122980530-0a3a9300-d399-11eb-9948-eb3985acd27f.png)
Should be:
![worldalign](https://user-images.githubusercontent.com/42101236/122980551-10307400-d399-11eb-8cac-c8dad96a1ade.png)

Relevant node definition:
```

local waterRange = 3

local freshWaterTint = { a = 128, r = 32, g = 40, b = 64 }
local seaWaterTint = { a = 200, r = 32, g = 64, b = 64 }

local seaWaterAlpha = 0
local freshWaterAlpha = 200

local waterGroups = { water = 1, liquid = 1 }

local animatedSeaWater = {
	{
		name = "mibi_seawater_a.png";
		animation = regularWaterAnimation;
		backface_culling = false;
		tileable_vertical=true;
		tileable_horizontal=true;
		align_style="world";
		scale = 4
	};
	{
		name = "mibi_seawater_a.png";
		animation = regularWaterAnimation;
		backface_culling = true;
		tileable_vertical=true;
		tileable_horizontal=true;
		align_style="world";
		scale = 4
	};
}

minetest.register_node( "core:seawater", {
	description = "Sea water",
	drawtype = "liquid",
	waving = 3;
	tiles = animatedSeaWater,
	special_tiles = animatedSeaWater,
	alpha = seaWaterAlpha,
	paramtype = ParamType.Light,
	walkable = false,
	pointable = false,
	diggable = false,
	buildable_to = true,
	is_ground_content = false,
	drop = "",		
	liquidtype = "source",
	liquid_alternative_flowing = "core:seawater_f",
	liquid_alternative_source = "core:seawater",
	liquid_viscosity = 1,
	liquid_renewable = false,
	liquid_range = waterRange,
	post_effect_color = seaWaterTint,
	groups = waterGroups
})
```

Wuzzy2 commented:
I really hope this PR gets more core dev attention.
For those who don't understand what this PR does:
Basically it fixes the broken depth sorting of many nodes, so you can actually start to use semitransparent nodes in your game glitch-free for the first time ever.


There is a huge amount of work put into this, it would be a shame if all the hours of hard work all went down the drain.

hecktest commented:
If the bug with world aligned textures is fixed and nobody reports any crippling performance issues, it's got my approval. Further optimizations are welcome but rendering correctness is a bigger priority, and alpha blended nodes are quite rare anyway. It does run reasonably well on an average PC.

x2048 commented:
@hecktest First take at world-aligned scaled textures, can you take a look?

hecktest commented:
Looks good now.
![aligned](https://user-images.githubusercontent.com/42101236/126800860-5b6f2ca2-948c-431e-ad1d-0a5179f236ad.png)

x2048 commented:
Hmmm.... I'll need to take a look. Is it the same texture?

hecktest commented:
Yes, I just toggled the "opaque water" setting.

x2048 commented:
@hecktest OK, so it was broken by the new solidness levels. Should be fixed now with new level for opaque liquids that re-enables fast face generation. Thanks for catching!

sfan5 commented:
FYI in case you haven't seen the Android build errors out: https://github.com/minetest/minetest/pull/11130/checks?check_run_id=3156128557

x2048 commented:
@sfan5 Fixed now, thanks for pointing.

hecktest commented:
Everything seems to work now. There could be some stutter if someone overuses alpha blending but you'd have to really abuse transparent nodes in order to make this noticeable. In any case, correct rendering comes first.

You could look into reordering indices instead of regenerating meshes if you have time, but if not I'll approve this regardless. (just fix the CI problems first)

x2048 commented:
@hecktest Can you elaborate about reordering indices? Right now the number of mesh buffers / meshes and their composition  depends on the player position to support fancy cases like transparent water surrounded by colored glass. Having two meshes and reordering vertices will not render correctly, or am I missing something? Thank you.

hecktest commented:
@x2048 
If we need to conditionally disable triangles for any reason, this can be done while sorting the index (element) buffer by simply not emitting some of the tris into the result. Having a few dead verts in your attribute buffers is a lot better than running the mesher again and recreating those entirely.
It's possible to do this, it would just require rethinking some of the data structures and the mesher process itself; you'd effectively defer and re-run the part of the mesher that decides which faces are actually emitted. (after modifying it so that it achieves this by not emitting elements while all the verts are actually present)
I do not know to what degree the mesher is already compatible with this approach.

This isn't terribly important though, who knows if we aren't going to redo meshing entirely after cleaning up the renderer.

x2048 commented:
@hecktest I've played with indicies, and yes, it will be much better to use them, but it's a topic for a different PR. I'll see if I get to writing a prosal anytime soon.

x2048 commented:
Just for transparency, I've rebased this branch on a reasonably resent master and extracted code for sorting mapblocks into #11651. What's left in this PR:
* Generating meshes with correct vertex order based on camera position, including many fixes for water / transparent liquids
* All the infrastructure around regenerating the meshes when the player moves
* New material solidness system to render leaves, glass and liquids more accurately.

Wuzzy2 commented:
Nice! I have tried it out briefly again, still works as before.

Andrey2470T commented:
> @Andrey2470T, I've pushed a fix for the flicker mentioned earlier, so this PR now fully fixes #10218.

I've decided to test now and I can confirm this fixes, thanks much. There is an another bug (that apparently the PR is not fixed yet) relates to nodes with mesh drawtype. If any of mesh faces uses alpha channel on a texture, then other parts of the same mesh using opaque parts of a texture and locating behind the first ones, they will not render like in that issue. It would be nice if it was fixed also.

x2048 commented:
@Andrey2470T See [here](https://github.com/x2048/minetest/tree/alpha_blend_indexed). Mostly works, but produces artifacts with certain arrangements of the triangles.

x2048 commented:
This pull request is largely superseded by #11696 and is now deprecated.
